### PR TITLE
Add override config layer for forced retry behavior

### LINKED ISSUE #108: Question: Short Circuiting Retriable While Testing Your App
The readme mentions being able to turn of retries in tests by doing the following:

```rb
Retriable.configure do |c|
  if Rails.env.test?
    c.tries = 1
end
```

But from what I can tell is only works when `Retriable.retriable` is itself called without the `tries` parameter. The default global config is being set, while the individual invocation parameters still take precedence. The same is probably true for all other parameters being set in this way.

Looking at the code this seems to indeed what is happening: https://github.com/kamui/retriable/blob/83aa4d47fe8c4551bcc2246236d9d1a7c9122730/lib/retriable.rb#L26

Is there something I'm missing? I could create my own wrapper function and check for the test environment there, but that is not what the readme suggests.

### PR DESCRIPTION
**Pull request by kamui**
## Summary

Adds `Retriable.override` and `Retriable.reset_override` APIs that force retry settings over local call options.

```ruby
Retriable.override(tries: 1, base_interval: 0)
```

### Precedence

```
override > local options > configure defaults
```

### Use case

Test short-circuiting: in test environments, `tries: 1` must win even when production code passes explicit local options like `Retriable.retriable(tries: 5)`. Overrides also force retry behavior over inherited custom `intervals` unless `intervals` is explicitly overridden.

### API

- `Retriable.override(opts)` accepts a non-empty hash of config keys to force globally.
- `Retriable.reset_override` clears the override.
- Contexts are supported: `Retriable.override(contexts: { api: { tries: 1 } })`.
- Invalid top-level keys and invalid context override keys raise `ArgumentError`, consistent with `Retriable.retriable`.

### Example: Rails test setup

```ruby
# config/initializers/retriable.rb
Retriable.configure do |c|
  c.tries = 3
  c.base_interval = 0.5
end

if Rails.env.test?
  Retriable.override(tries: 1, base_interval: 0, rand_factor: 0)
end
```

### Implementation

The override is stored as a sparse hash delta and merged lazily at each `retriable`/`with_context` call. Hash-valued options replace lower-precedence values rather than deep-merging, so forced options remain authoritative.

### Changes

- `lib/retriable.rb` — adds `override`, `reset_override`, validation, and supporting merge helpers
- `spec/retriable_spec.rb` — tests override precedence, contexts, edge cases, validation, and order isolation
- `README.md` — new Override section, updated Testing section with override examples
- `CHANGELOG.md` / `lib/retriable/version.rb` — version and changelog entries

Closes #108.

### CODE REVIEW

File: lib/retriable.rb [hunk 1]
```diff
@@ -128,13 +191,76 @@ def hash_exception_match?(exception, on, exception_list)
...
+                        override_contexts)
+    end
+    return {} if override_contexts.nil?
+
+    deep_dup(base_contexts)
+  end
+
+  def merged_context_options(contexts, context_key, options)
+    context_options = contexts[context_key].merge(options)
+
```
[the remark this answers is not in the document]
Comment (kamui) [reply lib/retriable.rb:195]: `merged_context_options` now validates that `contexts[context_key]` is a Hash before calling `.merge`, falling back to `{}` for nil/non-Hash values. Also tightened the override context options check to require `.is_a?(Hash)`. Added spec coverage for nil context values.

File: lib/retriable.rb [hunk 2]
```diff
@@ -128,13 +159,49 @@ def hash_exception_match?(exception, on, exception_list)
...
+      return deep_merge(base_contexts.is_a?(Hash) ? base_contexts : {}, override_contexts)
+    end
+    return {} if override_contexts.nil?
+
+    base_contexts
+  end
+
+  def merged_context_options(contexts, context_key, options)
+    context_options = contexts[context_key].merge(options)
+
```
[the remark this answers is not in the document]
Comment (kamui) [reply lib/retriable.rb:195]: Addressed together with the other `merged_context_options` comment. Both base context value and override context value are now validated as Hash before merging.

File: lib/retriable.rb [hunk 3]
```diff
@@ -16,19 +26,40 @@ def config
     @config ||= Config.new
   end
 
+  def override(opts = {})
+    opts.each_key do |k|
+      raise ArgumentError, "#{k} is not a valid option" unless Config::ATTRIBUTES.include?(k)
+    end
+    @override_config = opts.empty? ? nil : opts
+  end
```
[the remark this answers is not in the document]
Comment (kamui) [reply lib/retriable.rb:45]: `override` now deep-dups and freezes the provided opts hash so external mutation of the original hash cannot affect stored override behavior. Added a `deep_dup` private helper and spec coverage.

File: lib/retriable/version.rb [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # frozen_string_literal: true
 
 module Retriable
-  VERSION = "3.4.1"
+  VERSION = "3.4.0"
```
[the remark this answers is not in the document]
Comment (kamui) [reply lib/retriable/version.rb]: VERSION bumped to 3.4.2. The 3.4.1 release (monotonic clock, nil max_elapsed_time, dead float coercion) is already released, and the override feature is new functionality on top.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,14 +1,13 @@
...
-## 3.4.1
-
 - Fix: Use `Process.clock_gettime(CLOCK_MONOTONIC)` for elapsed time tracking so retry timing is immune to wall-clock adjustments (NTP, manual changes).
 - Fix: Handle `max_elapsed_time: nil` gracefully instead of raising `NoMethodError`.
 - Remove dead `* 1.0` float coercion in `ExponentialBackoff#randomize`.
 
 ## 3.4.0
 
 - Add `retry_if` option to support custom retry predicates, including checks against wrapped `exception.cause` values.
+- Add `override` and `reset_override` APIs to force retry settings over local call options when needed (for example, test short-circuiting).
```
[the remark this answers is not in the document]
Comment (kamui) [reply CHANGELOG.md:14]: CHANGELOG now has a separate 3.4.1 section for the already-released fixes, with the override feature under HEAD. Aligned with VERSION 3.4.2.

File: README.md [hunk 6]
```diff
@@ -390,7 +429,7 @@ Then in your test environment, you would need to set each context and the defaul
 
 ```ruby
 # spec/spec_helper.rb
-Retriable.configure do |c|
+Retriable.override do |c|
   c.multiplier    = 1.0
   c.rand_factor   = 0.0
   c.base_interval = 0
```
[the remark this answers is not in the document]
Comment (kamui) [reply README.md]: Updated the README per-context section. The wording now clarifies that top-level override values already take precedence over context-specific values, and per-context overrides are only needed for context-specific options like `:intervals` or `:on`.

File: lib/retriable.rb [hunk 7]
```diff
@@ -125,13 +190,89 @@ def hash_exception_match?(exception, on, exception_list)
...
+
+  def merged_contexts
+    contexts = deep_dup(config.contexts)
+    return contexts unless override_config&.key?(:contexts)
+
+    override_contexts = override_config[:contexts]
+    return deep_merge(contexts, override_contexts) if override_contexts.is_a?(Hash)
+    return {} if override_contexts.nil?
+
+    contexts
```
[the remark this answers is not in the document]
Comment (kamui) [reply lib/retriable.rb]: Already addressed in current code. `merged_contexts` returns `config.contexts` directly (line 167) when no override touches `:contexts`, avoiding the deep-dup allocation. When an override is present, base is normalized to `{}` for non-Hash values before `deep_merge`.

File: README.md [hunk 8]
```diff
@@ -142,6 +142,33 @@ Retriable.configure do |c|
...
+
+```
+override > local options > configure defaults
+```
+
+To clear an override:
+
+```ruby
+Retriable.reset_override
+```
```
[the remark this answers is not in the document]
Comment (kamui) [reply README.md:171]: Already documented in the Override section (README lines 162-165): "`#override` uses process-global state. Once set, it affects every caller and thread until `#reset_override` runs. Prefer setting it once at boot (or in test helpers), and avoid toggling it per request in multi-threaded runtimes."

File: lib/retriable.rb [hunk 9]
```diff
@@ -125,13 +188,72 @@ def hash_exception_match?(exception, on, exception_list)
...
   end
 
+  def override_config
+    @override_config
+  end
+
+  def merged_contexts
+    contexts = deep_dup(config.contexts)
+    return contexts unless override_config&.key?(:contexts)
+
```
[the remark this answers is not in the document]
Comment (kamui) [reply lib/retriable.rb]: Already addressed. `merged_contexts` has an early return at line 167 (`return config.contexts unless override_config&.key?(:contexts)`) that avoids the deep-dup when no override contexts are present.

File: README.md [hunk 10]
```diff
@@ -340,29 +367,31 @@ end
...
-  # ... default configuration
-
-  if Rails.env.test?
+if Rails.env.test?
+  Retriable.override do |c|
     c.tries = 1
+    c.base_interval = 0
+    c.rand_factor = 0
   end
 end
```
[the remark this answers is not in the document]
Comment (kamui) [reply README.md]: Already addressed. The Rails initializer example (README lines 373-385) now shows `Retriable.configure` for shared defaults followed by a conditional `Retriable.override` for test-only overrides.

File: lib/retriable.rb [hunk 11]
```diff
@@ -125,13 +188,76 @@ def hash_exception_match?(exception, on, exception_list)
...
+  def contexts_override_delta(base_contexts, override_contexts)
+    base_contexts = {} unless base_contexts.is_a?(Hash)
+    override_contexts = {} unless override_contexts.is_a?(Hash)
+
+    override_contexts.each_with_object({}) do |(context_key, override_value), delta|
+      next if base_contexts.key?(context_key) && base_contexts[context_key] == override_value
+
+      delta[context_key] = override_value
+    end
```
[the remark this answers is not in the document]
Comment (kamui) [reply lib/retriable.rb]: Not applicable to the current implementation. `override` now takes a plain hash (not a block), so `c.contexts.delete(:api)` is not possible through the override API. Context deletions would require assigning a full replacement hash via `Retriable.override(contexts: { ... })`, which works correctly.

File: lib/retriable.rb [hunk 12]
```diff
@@ -128,13 +159,49 @@ def hash_exception_match?(exception, on, exception_list)
...
+    return context_options unless override_config
+
+    override_contexts = override_config[:contexts]
+    return context_options unless override_contexts.is_a?(Hash)
+
+    override_context_options = override_contexts[context_key]
+    return context_options unless override_context_options
+
+    deep_merge(context_options, override_context_options)
+  end
```
[the remark this answers is not in the document]
Comment (kamui) [reply lib/retriable.rb:205]: Already addressed in the same commit (254b653). `merged_context_options` now validates both values with `.is_a?(Hash)` guards:

- `contexts[context_key]` → line 193: `base = {} unless base.is_a?(Hash)`
- `override_contexts[context_key]` → line 202: `return context_options unless override_context_options.is_a?(Hash)`

Non-Hash values (including `nil`) fall back to `{}` or skip the override merge entirely, so no `NoMethodError`/`TypeError` is possible.
